pubtools/_pulp/tasks/publish.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import sys
+import re
+from datetime import datetime
+from more_executors.futures import f_sequence
+
+from pubtools.pulplib import Criteria, Matcher, PublishOptions
+
+from pubtools._pulp.task import PulpTask
+from pubtools._pulp.cdn_cache import CDNCache
+from pubtools._pulp.services import PulpClientService, UdCacheClientService
+
+
+step = PulpTask.step
+
+LOG = logging.getLogger("pubtools.pulp")
+
+
+def publish_date(str_date):
+    # validates publish-before date
+    return datetime.strptime(str_date, "%Y-%m-%d")
+
+
+class Publish(PulpClientService, UdCacheClientService, PulpTask, CDNCache):
+    """Publish one or more Pulp repositories to the endpoints defined by their distributors.
+
+    This command will publish the Pulp repositories provided in the request or
+    fetched using the filters(url-regex or published-before) or an intersection
+    of input repositories and filters.
+    """
+
+    def add_args(self):
+        super(Publish, self).add_args()
+
+        self.parser.add_argument(
+            "--repo-ids", help="list of repos to be published", nargs="+"
+        )
+        self.parser.add_argument(
+            "--clean",
+            help="attempt to remove content not in the repo",
+            action="store_true",
+        )
+        self.parser.add_argument(
+            "--force", help="execute a full repo publish", action="store_true"
+        )
+        self.parser.add_argument(
+            "--published-before",
+            help="publish the repos last published before given date e.g. 2019-08-21",
+            type=publish_date,
+            default=None,
+        )
+        self.parser.add_argument(
+            "--repo-url-regex",
+            help="publish repos whose repo url match",
+            default=None,
+            type=re.compile,
+        )
+
+    def run(self):
+        to_await = []
+        LOG.debug("Begin publishing repositories")
+
+        # get repos applying filters
+        repos = self.get_repos()
+
+        # publish the repos found
+        publish_fs = self.publish(repos)
+
+        # wait for the publish to complete before
+        # flushing caches.
+        f_sequence(publish_fs).result()
+
+        # flush UD cache
+        to_await.extend(self.flush_ud(repos))
+
+        # flush CDN cache
+        to_await.extend(self.flush_cdn(repos))
```

Comment:
Agree. That's how it's in current implementation too. However, it's otherwise in clear-repo(which is what was referred). Have changed the sequence as suggested.

Reply:
CDN flush isn't really first with this update either because you never await completion of these CDN flushes before proceeding to UD.  I think that's fine though, note that we only flush cache for repomd.xml files which don't affect operation of UD at all (which only links directly to files under origin).